src/main/scala/com/microsoft/azure/cosmosdb/spark/rdd/CosmosDBRDDIterator.scala

```diff
@@ -360,62 +354,62 @@ class CosmosDBRDDIterator(hadoopConfig: mutable.Map[String, String],
+                    s"Retry Delay (ms): $retryDelayInMs")
+                  Thread.sleep(retryDelayInMs)
+                  retryCount += 1
+                } else {
+                  logError("Exhausted all retries on Service Unavailable exception")
+                  throw ex
+                }
+              case _ => ex.getCause match {
+                case docEx: DocumentClientException =>
+                  handleGoneException(connection, docEx)
+                  isGone = true
+                case _ =>
+                  logError(s"An IllegalStateException was thrown: ${ex.getMessage}")
               }
-            } else if (ex.getCause.isInstanceOf[DocumentClientException]) {
-              val docex = ex.getCause.asInstanceOf[DocumentClientException]
-              handleGoneException(connection, docex);
-              isGone = true
-            } else {
-              logError(s"An IllegalStateException was thrown: ${ex.getMessage}")
             }
-          case ex: Throwable => {
+          case ex: Throwable =>
             logError(s"UNHANDLED EXCEPTION: ${ex.getMessage}")
             throw ex
-          }
         }
       }
       iteratorDocument
     }
 
-    taskContext.addTaskFailureListener((context: TaskContext, ex: Throwable) => {
+    taskContext.addTaskFailureListener((_: TaskContext, ex: Throwable) => {
       logError("Handling Task Failure")
-      if (ex.isInstanceOf[DocumentClientException]) {
-        val dcx: DocumentClientException = ex.asInstanceOf[DocumentClientException]
-        if (dcx.getStatusCode == StatusCodes.SERVICE_UNAVAILABLE) {
-          logError("Service Unavailable")
-          connection.reinitializeClient(shouldReinitializeThroughput = false)
-        }
-      } else if (ex.isInstanceOf[IllegalStateException]
-        && ex.getCause != null
-        && ex.getCause.isInstanceOf[DocumentClientException]){
-        logError("Illegal State Exception with Service Unavailable")
-        val dcx: DocumentClientException = ex.getCause.asInstanceOf[DocumentClientException]
-        if (dcx.getStatusCode == StatusCodes.SERVICE_UNAVAILABLE) {
-          connection.reinitializeClient(shouldReinitializeThroughput = false)
-        }
+      ex match {
+        case dcx: DocumentClientException =>
+          if (dcx.getStatusCode == StatusCodes.SERVICE_UNAVAILABLE) {
+            logError("Service Unavailable")
+            connection.reinitializeClient()
+          }
+        case _: IllegalStateException if ex.getCause.isInstanceOf[DocumentClientException] && ex.getCause != null =>
+          logError("Illegal State Exception with Service Unavailable")
+          val dcx: DocumentClientException = ex.getCause.asInstanceOf[DocumentClientException]
+          if (dcx.getStatusCode == StatusCodes.SERVICE_UNAVAILABLE) {
+            connection.reinitializeClient()
+          }
+        case _ =>
```

Comment:
I'm not sure what the desired behaviour here is but we should log something if we're going to all exceptions.